libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/builder/LocationProvider.kt

```diff
@@ -0,0 +1,66 @@
+package com.mapbox.navigation.core.routeoptions.builder
+
+import android.location.Location
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.core.trip.session.LocationMatcherResult
+import com.mapbox.navigation.core.trip.session.LocationObserver
+import com.mapbox.navigation.core.trip.session.TripSessionLocationProvider
+import com.mapbox.navigation.utils.internal.toPoint
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.yield
+import kotlin.coroutines.resume
+
+internal interface LocationProvider {
+    suspend fun getCurrentLocation(): CurrentLocation
+}
+
+internal data class CurrentLocation(
+    val point: Point,
+    val bearing: Double?,
+    val zLevel: Int?
+)
+
+internal class LocationFromTripSessionProvider(
```

Comment:
Thank you. I agree it can. I'm not sure I understand why an extension function is better than a wrapper class. Please share your thoughts about _exception function vs wrapper class_ .

Not sure if the provided code is expected to be fully functional but it will crash on `unregisterLocationObserver(this)`. You can reed and share your opinion about the crash [here](https://github.com/mapbox/mapbox-navigation-android/pull/5427#discussion_r801856654)

Reply:
> Please share your thoughts about exception function vs wrapper class .

There is nothing wrong with _wrapper classes_. They are extremely useful in languages that doesn't provide any other mechanism to add new functionality to an existing class or when we need to store some state.

In this specific case, we can extend TripSessionLocationProvider functionality by utilizing Kotlin extensions.

> Not sure if the provided code is expected to be fully functional but it will crash on unregisterLocationObserver(this). 

I am assuming the crash is caused by ConcurrentModificationException. We should create new ticket for patching `unregisterLocationObserver()`.

Reply:
If I replace wrapper by an extension function then `NavRouteOptionsBuilder` will start to depend on the `TripSessionLocationProvider` directly. But I would prefer `NavRouteOptionsBuilder` to work with an interface `LocationProvider` which can be implemented by other classes.